pydatastructs/trees/m_ary_trees.py

```diff
@@ -0,0 +1,168 @@
+from pydatastructs.utils import MAryTreeNode
+from pydatastructs.linear_data_structures.arrays import ArrayForTrees
+
+__all__ = [
+    'MAryTree'
+]
+
+class MAryTree(object):
+    """
+    Abstract m-ary tree.
+
+    Parameters
+    ==========
+
+    max_children
+        Optional, specifies the maximum number of children
+        a node can have. Defaults to 2 in case nothing is
+        specified.
+    root_data
+        Optional, the root node of the binary tree.
+        If not of type MAryTreeNode, it will consider
+        root as data and a new root node will
+        be created.
+    key
+        Required if tree is to be instantiated with
+        root otherwise not needed.
+    comp: lambda
+        Optional, A lambda function which will be used
+        for comparison of keys. Should return a
+        bool value. By default it implements less
+        than operator.
+    is_order_statistic: bool
+        Set it to True, if you want to use the
+        order statistic features of the tree.
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/M-ary_tree
+    """
+
+    __slots__ = ['root_idx', 'max_children', 'comparator', 'tree', 'size',
+                 'is_order_statistic']
+
+
+    def __new__(cls, max_children=2, key=None, root_data=None, comp=None,
+                is_order_statistic=False):
+        obj = object.__new__(cls)
+        if key is None and root_data is not None:
+            raise ValueError('Key required.')
+        key = None if root_data is None else key
+        root = MAryTreeNode(key, root_data)
+        root.is_root = True
+        obj.root_idx = 0
+        obj.max_children = max_children
+        obj.tree, obj.size = ArrayForTrees(MAryTreeNode, [root]), 1
+        obj.comparator = lambda key1, key2: key1 < key2 \
+                        if comp is None else comp
+        obj.is_order_statistic = is_order_statistic
+        return obj
+
+    def insert(self, key, data):
+        """
+        Inserts data by the passed key using iterative
+        algorithm.
+
+        Parameters
+        ==========
+
+        key
+            The key for comparison.
+        data
+            The data to be inserted.
+
+        Returns
+        =======
+
+        None
+        """
+        raise NotImplementedError("This is an abstract method.")
+
+    def delete(self, key, **kwargs):
+        """
+        Deletes the data with the passed key
+        using iterative algorithm.
+
+        Parameters
+        ==========
+
+        key
+            The key of the node which is
+            to be deleted.
+        balancing_info: bool
+            Optional, by default, False
+            The information needed for updating
+            the tree is returned if this parameter
+            is set to True. It is not meant for
+            user facing APIs.
```

Comment:
```suggestion
```
The above parameter was added for some different reasons in `BinaryTree`. Let's keep the API for `MAryTree` minimal for now, if needed we can add this later on.

Reply:
Fixed.